File: Week11/Week11_ProblemC.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

class Node {
	int iq, weight;
	Node (int iq, int weight) {
		this.iq = iq;
		this.weight = weight;
	}
}

public class Week11_ProblemC {
	
	static int n;
	static Node[] nodes;
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		//System.setIn(new FileInputStream("2.in"));	
	
		InputStreamReader r = new InputStreamReader ( System . in );
		BufferedReader in = new BufferedReader ( r );
		int testCasesCount = Integer.parseInt(in.readLine());
		String[] tokens;
		
		for (int i = 1; i <= testCasesCount; i++) {
			n = Integer.parseInt(in.readLine());
			
			nodes = new Node[n];
			
			for (int j = 0; j < n; j++) {
				tokens = in.readLine().split(" ");
				nodes[j] = new Node(Integer.parseInt(tokens[0]), Integer.parseInt(tokens[1]));
			}
			
			System.out.println(buildResult(i));
			
			if (i != testCasesCount)
				in.readLine();
		}
		
	}
	
	public static StringBuilder buildResult(int caseNo) {		
		
		Arrays.sort(nodes, (a, b) -> { int res = Integer.compare(a.iq, b.iq); return res == 0 ? Integer.compare(b.weight, a.weight) : res; });
		
		int[] currSum = new int[nodes.length];
		currSum[0] = 1;
		int maxSum = 1;
		
//		for (int x = 0; x < nodes.length; x++) {
//			System.out.println(nodes[x].iq + "\t" + nodes[x].weight);
//		}
		
		for(int x = 1; x < nodes.length; x++) {
			currSum[x] = 1;
			
			for (int y = 0; y < x; y++) {
				if (nodes[x].weight < nodes[y].weight && currSum[x] < currSum[y] + 1) {
					currSum[x] = currSum[y] + 1;
					
					if (currSum[x] > maxSum) {
						maxSum = currSum[x];
						break;
					}
				}
			}
		}
		
		StringBuilder sb = new StringBuilder(12);
		sb.append("Case #");
		sb.append(caseNo);
		sb.append(": ");
		sb.append(maxSum);
		
		return sb;
	}
}

File: Week11/Week11_ProblemC.cpp
//
//  main.cpp
//  Week11_ProblemC
//
//  Created by Zain Ul Abideen on 23/01/2017.
//  Copyright © 2017 Zain Ul Abideen. All rights reserved.
//

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

void buildResult(int i);

struct Node {
    int iq, weight;
};

int n;
vector<Node> nodes;

int main(int argc, const char * argv[]) {
    ios_base :: sync_with_stdio ( false );
    
    int testCases;
    
    cin >> testCases;
    
    for (int i = 1; i <= testCases; i++) {
        cin >> n;
        nodes.resize(n);
        
        for (int j = 0; j < n; j++) {
            nodes[j] = Node();
            cin >> nodes[j].iq;
            cin >> nodes[j].weight;
        }
        
        buildResult(i);
        
//        if (i != testCases)
//            in.readLine();
    }
    
    return 0;
}

void buildResult(int caseNo) {
    
    sort(nodes.begin(), nodes.end(), [](Node a, Node b) { return a.iq == b.iq ? b.weight < a.weight : a.iq < b.iq; });
    
    int* currSum = new int[nodes.size()];
    currSum[0] = 1;
    int maxSum = 1;
    
    //		for (int x = 0; x < nodes.length; x++) {
    //			System.out.println(nodes[x].iq + "\t" + nodes[x].weight);
    //		}
    
    int size = nodes.size();
    
    for(int x = 1; x < size; x++) {
        currSum[x] = 1;
        
        for (int y = 0; y < x; y++) {
            if (nodes[x].weight < nodes[y].weight && currSum[x] < currSum[y] + 1) {
                currSum[x] = currSum[y] + 1;
                
                if (currSum[x] > maxSum) {
                    maxSum = currSum[x];
                    break;
                }
            }
        }
    }
    
    cout << "Case #" << caseNo << ": " << maxSum << endl;
}
